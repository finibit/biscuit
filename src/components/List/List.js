import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import shortid from 'shortid'
import styles from '../../themes'
import ListItem from './parts/ListItem'

const ListStyled = styled.div`
	${styles.margin}
	${styles.border}

	padding: 0;
	display: flex;
	flex-direction: column;
	border-top-left-radius: ${(props) => props.theme.global.borders[0].radius};
	border-top-right-radius: ${(props) => props.theme.global.borders[0].radius};
	border-bottom-left-radius: ${(props) => props.theme.global.borders[0].radius};
	border-bottom-right-radius: ${(props) => props.theme.global.borders[0].radius};
`

/** A list of items. */
const List = (props) => {
	const {
		items,
		renderItem,
		margin,
		themeElement,
		...rest
	} = props

	return (
		<ListStyled
			$element={themeElement}
			$margin={margin}
			$border={0}
			{...rest}
		>
			{items.map((item) => {
				const elem = renderItem(item)

				if (elem.key != null) {
					return elem
				}

				return React.cloneElement(elem, { key: shortid() })
			})}
		</ListStyled>
	)
}

List.propTypes = {
	/** Array of items. */
	items: PropTypes.arrayOf(PropTypes.any),

	/** Item renderer. */
	renderItem: PropTypes.func,

	/** The amount of margin around the list. */
	margin: PropTypes.oneOfType([
		PropTypes.number,
		PropTypes.string,
		PropTypes.object,
		PropTypes.array,
	]),

	/** Theme element. */
	themeElement: PropTypes.string,
}

List.Item = ListItem

List.defaultProps = {
	items: [],
	renderItem: (item) => (<List.Item>{item}</List.Item>),
	margin: 'none',
	themeElement: 'List',
}

export default List
