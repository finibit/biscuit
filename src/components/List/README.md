### Usage

#### Items

```jsx
<List
  items={['Red', 'Green', 'Blue']}
/>
```

#### Margin

```jsx
<List
  items={['Red', 'Green', 'Blue']}
  margin={3}
/>
```
