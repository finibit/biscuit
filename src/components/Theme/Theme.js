import React from 'react'
import PropTypes from 'prop-types'
import { ThemeProvider, createGlobalStyle } from 'styled-components'
import { mergeTheme, baseTheme, themeGet } from '../../themes'
import types from '../../types'

const GlobalStyle = createGlobalStyle`
	html, body {
		width: 100%;
		height: 100%;
		margin: 0;
		padding: 0;
	}

	body {
		color: ${themeGet.color('black', 'inherit')};
		font-family: ${themeGet.fontFamily('secondary')};
		-webkit-font-smoothing: antialiased;
	}
`

const Theme = (props) => {
	const {
		theme,
		global,
	} = props

	return (
		<ThemeProvider theme={mergeTheme(baseTheme, theme)}>
			<div>
				{ global && <GlobalStyle /> }
				{props.children}
			</div>
		</ThemeProvider>
	)
}

Theme.propTypes = {
	global: PropTypes.bool,
	...types.theme,
}

Theme.defaultProps = {
	global: true,
	theme: baseTheme,
}

export default Theme
