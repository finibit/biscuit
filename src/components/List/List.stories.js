import React from 'react'
import { storiesOf } from '@storybook/react'
import Wrapper from '../../Wrapper'
import List from './List'

storiesOf('List', module)
  .addDecorator(Wrapper)
  .add('with items', () => (
    <List items={[
      "Invent the universe rich in mystery are creatures of the cosmos (...)",
      "Rogue tingling of the spine another world tesseract at the edge of forever culture?",
      "Of brilliant syntheses cosmos vanquish the impossible white dwarf Rig Veda (...)",
      "The sky calls to us dream of the mind's eye hundreds of thousands (...)"
    ]} />
  ))