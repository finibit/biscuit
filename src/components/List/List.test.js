import React from 'react'
import renderer from 'react-test-renderer'
import 'jest-styled-components'

import Theme from '../Theme'
import List from './List'

describe('List', () => {
  test('empty', () => {
    const component = renderer.create((
      <Theme>
        <List items={[]} />
      </Theme>
    ))
    expect(component.toJSON()).toMatchSnapshot()
  })

  test('with items', () => {
    const component = renderer.create((
      <Theme>
        <List items={[
          'Item 1',
          'Item 2',
          'Item 3'
        ]}
        />
      </Theme>
    ))
    expect(component.toJSON()).toMatchSnapshot()
  })
})
