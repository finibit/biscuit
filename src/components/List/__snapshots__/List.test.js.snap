// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List renders empty 1`] = `
.c0 {
  margin: 0;
  border: 1px solid #d8d8d8;
  padding: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  border-top-left-radius: 0.40187757201646096rem;
  border-top-right-radius: 0.40187757201646096rem;
  border-bottom-left-radius: 0.40187757201646096rem;
  border-bottom-right-radius: 0.40187757201646096rem;
}

<div>
  <div
    className="c0"
  />
</div>
`;

exports[`List renders items custom 1`] = `
.c1 {
  padding: 0.694rem 1.440rem 0.694rem 1.440rem;
  border: 1px solid #d8d8d8;
  margin: 0;
  border-left: none;
  border-top: none;
  border-right: none;
}

.c1:last-child {
  border-bottom: none;
}

.c0 {
  margin: 0;
  border: 1px solid #d8d8d8;
  padding: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  border-top-left-radius: 0.40187757201646096rem;
  border-top-right-radius: 0.40187757201646096rem;
  border-bottom-left-radius: 0.40187757201646096rem;
  border-bottom-right-radius: 0.40187757201646096rem;
}

<div>
  <div
    className="c0"
  >
    <div
      className="c1"
    >
      Custom Red
    </div>
    <div
      className="c1"
    >
      Custom Green
    </div>
    <div
      className="c1"
    >
      Custom Blue
    </div>
  </div>
</div>
`;

exports[`List renders items default 1`] = `
.c1 {
  padding: 0.694rem 1.440rem 0.694rem 1.440rem;
  border: 1px solid #d8d8d8;
  margin: 0;
  border-left: none;
  border-top: none;
  border-right: none;
}

.c1:last-child {
  border-bottom: none;
}

.c0 {
  margin: 0;
  border: 1px solid #d8d8d8;
  padding: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  border-top-left-radius: 0.40187757201646096rem;
  border-top-right-radius: 0.40187757201646096rem;
  border-bottom-left-radius: 0.40187757201646096rem;
  border-bottom-right-radius: 0.40187757201646096rem;
}

<div>
  <div
    className="c0"
  >
    <div
      className="c1"
    >
      Red
    </div>
    <div
      className="c1"
    >
      Green
    </div>
    <div
      className="c1"
    >
      Blue
    </div>
  </div>
</div>
`;

exports[`List renders with custom margin 1`] = `
.c0 {
  margin: 1.440rem;
  border: 1px solid #d8d8d8;
  padding: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  border-top-left-radius: 0.40187757201646096rem;
  border-top-right-radius: 0.40187757201646096rem;
  border-bottom-left-radius: 0.40187757201646096rem;
  border-bottom-right-radius: 0.40187757201646096rem;
}

<div>
  <div
    className="c0"
  />
</div>
`;
